android/src/main/java/io/ably/flutter/plugin/AblyFlutterPlugin.java

```diff
@@ -1,28 +1,43 @@
 package io.ably.flutter.plugin;
 
+import android.app.Activity;
 import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+import android.os.Handler;
 
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 
-import java.lang.reflect.Method;
+import com.google.firebase.messaging.RemoteMessage;
 
+import java.nio.channels.CompletionHandler;
+
+import io.ably.flutter.plugin.generated.PlatformConstants;
+import io.ably.flutter.plugin.push.CompletionHandlerWithRemoteMessage;
 import io.ably.flutter.plugin.push.PushActivationEventHandlers;
-import io.flutter.Log;
+import io.ably.flutter.plugin.push.PushMessagingEventHandlers;
 import io.flutter.embedding.engine.plugins.FlutterPlugin;
+import io.flutter.embedding.engine.plugins.activity.ActivityAware;
+import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
 import io.flutter.plugin.common.BinaryMessenger;
 import io.flutter.plugin.common.MethodChannel;
 import io.flutter.plugin.common.MethodCodec;
+import io.flutter.plugin.common.PluginRegistry;
 import io.flutter.plugin.common.PluginRegistry.Registrar;
 import io.flutter.plugin.common.StandardMethodCodec;
 
-public class AblyFlutterPlugin implements FlutterPlugin {
+public class AblyFlutterPlugin implements FlutterPlugin, ActivityAware, PluginRegistry.NewIntentListener {
+    public static Boolean isActivityRunning = false;
```

Comment:
From reading the code on Github, it may not be clear, but in the code, `isActivityRunning` (now `isMainActivityRunning`) is used in `FirebaseMessagingReceiver`, which is a statically declared (launched even when the app is off) broadcast receiver. Because there is no guarantee that MainActivity/ FlutterActivity is running when this broadcast receiver is launched (and by extension the user's Dart code), the broadcast receiver **needs to know if it should launch the user's dart application**. This is documented in `FirebaseMessagingReceiver`'s `onReceive` method:
```java
    } else if (AblyFlutterPlugin.isMainActivityRunning) {
      // Flutter is already running, just send a background message to it.
      Intent onMessageReceivedIntent = new Intent(PUSH_ON_BACKGROUND_MESSAGE_RECEIVED);
      onMessageReceivedIntent.putExtras(intent.getExtras());
      LocalBroadcastManager.getInstance(context).sendBroadcast(onMessageReceivedIntent);
    } else {
```

# Another problem
In theory though, there can be many AblyFlutterPlugins running, so **just** **a boolean** is not good here. This case only happens when a user creates another FlutterActivity or FlutterFragment which (automatically if the user doesn't configure it) registers new plugin instances, including a new AblyFlutterPlugin. This would happen if a user is using Flutter in their Android app ("add-to-app" use cases).

Instead, we'd need to keep track of the counts of FlutterEngines. If it is 0, create one, otherwise do not. Each AblyFlutterPlugin will increment/ decrement this counter inside the `ActivityAware` interface methods, specifically: 

Or more simply to remove this static field, always create a new FlutterEngine if the app is the background, even if the app is already running. This would be slightly wasteful. 

TLDR: WDYT? 
- Add complexity to add support for add-to-app (multiple FlutterEngines)
- Remove complexity whilst adding support by always launching a FlutterEngine, even when not needed
- Keep this static field, and do not support add-to-app

Reply:
If I'm reading your responses correctly then I think:

1. Yes, "Keep this static field, and do not support add-to-app"
2. Also, "create an issue to add support for Add-to-app (very low priority)"

Reply:
@ben-xD feel free to resolve this conversation on my behalf once you're satisfied that it's settled - ideally adding a link to the issue you create here for the record.

Reply:
Cool, here it is: https://github.com/ably/ably-flutter/issues/183